### ISSUE TITLE
proposal: modules

### ISSUE THREAD

**Issue by tehn** [monome/norns#1384]
aka mods (modifications?)

basically, libs that are user-selectable to be "always loaded" at script init

existing cases:
- `passthrough` midi util
- `n160`
- more grid options ie rotation and brightness

rather than extend the menu system more and more, i'd like to propose a simple system whereupon libraries can arbitrarily add functionality (with UI in the param menu) or redefine existing behavior (in a non-breaking way... such as adding extra midi functions).

this is pretty easy to do by modifying `norns/lua/core/script.lua`

```lua
Script.init = function()
  print("# script init")
  params.name = norns.state.shortname
  -- lines below added to always load this specific lib:
  if util.file_exists(_path.code.."passthrough") then
    local _passthrough = include 'passthrough/lib/passthrough'
    _passthrough.init()
  end
  -- continue
  init()
  _norns.screen_save()
end
```

rather than a single file, this section could call a second file which has a list of activated modules/libs. editing this file (say in `dust/data/system.modules` or something) could change what gets loaded

but, in the marketplace/gallery style of community collaboration, it'd probably be preferable to have a menu UI for scanning all available mods, with a list of enable/disable.

potential issues:

- namespace collisions (we need some conventions)
- init behavior
- pre-existing libs: ie, a script that already `includes` or `requires` a mod need not be re-loaded
- general complexity
- param menu overload. (this is in fact not really a issue to me, as this stuff should all be optional. what would be overload is a bigger and bigger menu/param system by default.)

all comments and collaborators welcome.

**Comment by nattog** [comment]
I think its an elegant solution to handling a growing ecosystem of libraries. Some initial questions and thoughts:

- I think this functionality would benefit from some form of menu UI to be accessible and reduce complexity in use/user error. 
- In which case, how to distinguish between modules and scripts? Is some form of flag or tag needed? 
- Could this be controllable on individual scripts or applied for all scripts?
- Does it require some degree of conformity/shared standards across libs?
- Taking Passthrough as an example, it can interact with a script by defining callback functions at script level called from MIDI events handled within Passthrough. Will this proposal permit this behaviour - presumably there won't be any scoping issues?

**Comment by jeffbr13** [comment]
One consideration for symmetry with initialisation: module **teardowns/cleanup**.

Correct me if I’m wrong, but if this module system provides a  recommended point to hook arbitrary functionality for a whole norns system then a door is opened for many possibilities: e.g. a user’s custom audio routing set globally or only for certain scripts. https://github.com/xmacex/connect-opz/issues/5 could be another example use-case! 

Some modules would likely require a mirroring of their initialisation if they, for example, modify OS-wide state like udev or amidi rules. Without a place to perform the proper incantations to revert these upon exit from a script, modifications of the system state will build up over time as you enter and exit further scripts. Configurations could become hairy and error-prone, and only fixable with a reboot. 

Forgive me for unfamiliarity with Lua, but Python’s context managers provide one inspiration for implementation: <https://book.pythontips.com/en/latest/context_managers.html>. 
They are commonly used to open/close files, network connections, or wrap any functionality in a temporary context which you want to disappear when finished. Upon entering a context some setup code is run. Upon exit of the context some tear-down code is run.
They are nestable. The first context you enter has its exit code called last. Like a Matryoshka doll. 🪆
I am unsure how this pattern would translate into Lua, but I believe in other languages it is often implemented with either `try`/`finally` blocks or function callbacks.

**Comment by tehn** [comment]
> how to distinguish between modules and scripts? Is some form of flag or tag needed?

firstly, should "mods" just be "libs"? if yes, they're just lua files inside a scripts' `/lib` folder, which is hidden from the SELECT menu if no, we could support a `/mod` folder within scripts.

but this is a broader topic regarding how we differentiate requirements:
- if a script _requires_ a lib (think `fourwaves.lua` which is just a simple softcut sampler) which isn't really a good candidate for universal conditional inclusion via a menu (this is absolutely the case with fourwaves as it redefines softcut behavior and would break other softcut scripts)... this to me isn't in the spirit of a "module" --- even though this lib could be just as well used/included in another script (which the other script should do explicitly... this is how the include system works now)
- a "module" should be something that perhaps isn't absolutely required by a script. think add-on, such as previous examples of grid-rotation or midi passthru (note, i haven't looked at passthrough... but my abstract idea is that it should redefine global midi handlers and add some option params to route midi data).
- there will likely be conditions (i can't think of why/how, but figure it's inevitable) that certain scripts will _require_ a mod, because it wants to control it programmatically. we need to handle duplicate loading in the case of `include 'mod/whatever'` if the mod is also listed in the global-include list (via menu/etc)

> Could this be controllable on individual scripts or applied for all scripts?

i'd thought of this and it again becomes a UI issue: as if every script needs its own config. which in my mind is overly complex and really worth considering some example use cases.

> Does it require some degree of conformity/shared standards across libs?

there need to be defined init/cleanup syntax that can be returned/collected from the script.init function, so they can be stored and called appropriately. this shouldn't be too weird ie at the end of a mod:

```
return {init_func, cleanup_func}
```

_or_ we could pack all of the mod functionality into a subtable `mod`. think `mod.passthru` and `mod.grid_rota` tables


> Taking Passthrough as an example, it can interact with a script by defining callback functions at script level called from MIDI events handled within Passthrough. Will this proposal permit this behaviour - presumably there won't be any scoping issues?

if we do a subtable, the script could check for existence of `mod.passthu` and then make further definitions.

> One consideration for symmetry with initialisation: module teardowns/cleanup.

yes! right now the global Lua environment is cleared and restored between script runs, so script cleanup functions are largely unused (because tables/etc get GC'd)--- so `cleanup` is more used for writing data to disk/etc at close... it's more of a "all done" callback.

> Some modules would likely require a mirroring of their initialisation if they, for example, modify OS-wide state like udev or amidi rules. Without a place to perform the proper incantations to revert these upon exit from a script, modifications of the system state will build up over time as you enter and exit further scripts. Configurations could become hairy and error-prone, and only fixable with a reboot.

ugh. yes, anything which modifies the system outside of lua would need a reversing option.

addressed earlier, having an init/cleanup per module would hopefully address this.

---

parallel idea is having modules which change the menu system itself, ie, a different LEVELS or TAPE screen.

**Comment by jeffbr13** [comment]
Another use-case:
- [Midigrid: use launchpads & midi grid controllers with Norns (Update 2021-05-19) - Library - lines](https://llllllll.co/t/midigrid-use-launchpads-midi-grid-controllers-with-norns-update-2021-05-19/42336)
  - https://github.com/jaggednz/midigrid
  - https://norns.community/authors/jaggednz/midigrid

Used as a library to provide grid-compatibility with other controllers.

When this context is wrapped around a Script, the Script and any child contexts should be able use the configured controller as if it were a Monome Grid without knowing that it’s not a real Grid.

**Comment by jeffbr13** [comment]
> but this is a broader topic regarding how we differentiate requirements:
> 
> * if a script _requires_ a lib […] which isn't really a good candidate for universal conditional inclusion via a menu […]

Indeed, it seems these are the shared 'libraries' of functionality as in programming. Libraries aren't _usually_ end-user configurable: the calling application/script may configure and use a lib however it likes. The intention is merely deduplication of source code for common uses. Although there exist 'system libraries' which have out-of-band configuration but they can be a headache to program and test against (e.g. OpenSSL and it's .

Is the new idea here that 'scripts may declare that they depend upon library X @ version Y'? Does norns already use a Lua package manager like JS's NPM or Python's Pip? If not, bringing one in may solve some of those issues, at the cost of being itself a dependency and bringing requirements of its own.

> we need to handle duplicate loading in the case of `include 'mod/whatever'` if the mod is also listed in the global-include list (via menu/etc)

Avoiding duplicated dependencies (à la the [NPM/`node_modules` problem](https://npm.github.io/how-npm-works-docs/npm3/duplication.html)) will be something to avoid with libraries for certain.
 
> * a "module" should be something that perhaps isn't absolutely required by a script […]

> parallel idea is having modules which change the menu system itself, ie, a different LEVELS or TAPE screen.

This may be the new (second) concept: a mod/component/plugin which potentially provides GUI elements for config? 

Now I'm wondering how the different concepts of [plugins](https://en.wikipedia.org/wiki/Plug-in_%28computing%29) vs [pluggable components](https://martinfowler.com/bliki/SoftwareComponent.html) vs [dependency-injection](https://martinfowler.com/articles/injection.html) compose in a software system, and which of them might be more suitable for norns and Lua.

I wonder, should these plugin mods either:
1. simply 'wrap around' a script, to do something before and after the script to modify the running system in-place? Or
2. pass something (themselves? functions?) down into the scripts, dependency-injection style?

(1) sounds simpler, but doesn't speak to shared libraries. (2) sounds more in tune with the combined libs/plugin mods/components idea.

**Comment by ngwese** [comment]
handful of observations / suggestions. there are two lifetimes for lua "state" in norns right now. one is the lifetime of a script, the other is the lifetime of `matron` itself. my understanding of the `passthrough`, `n160`, and `midigrid` libraries is that they want to _extend_ the ambient functionality provided by `matron` for scripts - they want to live as long as matron lives. as of today there is no mechanism for extending `matron` other than a script explicitly loading one of these libraries which is inconvenient because it requires either the script author or script user to modify the script code. the behavior we're after is that a user of say `midigrid` would  just express their desire to use it and neither the scripts they use nor the authors of those scripts need to explicitly accommodate that. so i would argue these "extensions":

- ...will generally work through side-effects, modifying aspects of matron
- ...are adding ambient functionality therefore it makes sense to load them when matron starts (as opposed to when a script starts)

ui wise it seems reasonable to imagine something (say in the SYSTEM menu) which allows one to select the set of extensions / modifications (mods) to be loaded when matron starts. if/when one wants to disable an extension they deselect it and restart matron. within `startup.lua` (i think) it would be easy enough to `require` each extension and catch errors to ensure we at least make it to the menu system.

an additional benefit of thinking about these as extensions tied matron's lifetime is that should an extension both tweak matron and provide additional api for use in a script it may likely be sufficient to just use `require` as normal within scripts to access the api. the extension code wouldn't be redundantly loaded by scripts because lua would just return the module previously loaded by matron at startup.

so what about the case where the extension wants to provide a consistent initial state when a script starts? i would propose adding something like a `hook` api to matron itself. the extension could register one of its functions to be called by the appropriate hooks such as `script_pre_init` or `script_post_cleanup`. matron's `Script.init` implementation would in turn would invoke any registered callback. in practice i think an extension would look like this (pseudocode):

```lua
local hook = require('hook')

local extension_api = {}
local extension_state = {}

function extension_api.foo() ... end
function extension_api.bar() ... end

local function __init()
 ...
end
hook.script_pre_init.register('midigrid-init', __init)

local function __cleanup()
  ...
end
hook.script_post_cleanup.register('midigrid-cleanup', __cleanup)

local function __setup()
  ...
end
hook.system_post_startup.register('midigrid-setup', __setup)


return extension_api
```

one aspect of this approach that i think is particularly valuable is that it makes it more clear that an extension is not just a library, it is a piece of code which is specifically written to extend the behavior of matron and it does this by implementing one or more hooks. the hooks pattern is extensible, if for example we found a case where an extension needed/wanted to augment the saving of system state - we create a new hook that passed the system state to be saved as a table to all registered callbacks for that hook ahead of saving state to disk.

regarding where/how to package and find these extensions one option (similar to proposals above) would be convention. my suggestion would be to just use `dust/code/<script_dir>/lib/extension.lua`, essentially allowing each installed project/script to offer a single (optional) extension. the <script_dir> i.e. `passthrough` would dictate the name presented in the SYSTEM menu. 

i'd be happy to prototype/implement something like the above.

**Comment by tehn** [comment]
fixed by #1385
